### ISSUE TITLE
MacOS: 1. `pip install cdl` error - 2. Missing menus

### ISSUE THREAD

**Issue by marcel-goldschen-ohm** [Codra-Ingenierie-Informatique/DataLab#15]
`pip install cdl` in fresh python3 environment on MacOS 12.6.9 fails on install of PlotPy with error:
```
Building wheels for collected packages: PlotPy
  Building wheel for PlotPy (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for PlotPy (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [10 lines of output]
      ./src/scaler.cpp:58:35: error: a space is required between consecutive right angle brackets (use '> >')
      typedef XYTransform<Array1D<double>> XYScale;
                                        ^~
                                        > >
      ./src/scaler.cpp:616:56: error: a space is required between consecutive right angle brackets (use '> >')
          return scale_src_dst<Params, NoScale<ST, npy_uint32>>(p, scale);
                                                             ^~
                                                             > >
      2 errors generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for PlotPy
Failed to build PlotPy
ERROR: Could not build wheels for PlotPy, which is required to install pyproject.toml-based projects
```

Is this not the correct way to install DataLab on MacOS?

**Comment by PierreRaybaut** [comment]
Thanks for the feedback and for trying to play with DataLab!

Would you be able to try something to help me understand what's going on?

This would require to clone the [PlotPy repository](https://github.com/PlotPyStack/plotpy) and to slightly change the `setup.py` script by adding the following line just before line 55 (before the call to the `setup` function):
```python
CFLAGS_CPP += ["-std=c++11"]
````

And then try and build PlotPy from source:
```
python setup.py build
```
just to see if it works.

**Comment by marcel-goldschen-ohm** [comment]
This seems to have worked. At least I get no errors during the build.

**Comment by PierreRaybaut** [comment]
Great, thanks for the feedback.
I'll try to release PlotPy v2.0.3 as quickly as possible with this change.

In the meantime, if you type:
```
python setup.py build install
```
...you may retry installing DataLab with `pip install cdl`.

**Comment by marcel-goldschen-ohm** [comment]
Thanks for the quick fix @PierreRaybaut !!!

**Comment by marcel-goldschen-ohm** [comment]
I was able to run `cdl` from command line and I get the UI window. Not sure if it really matters but FYI, I also get a lot of warnings stating `ModuleNotFoundError: No module named 'plotpy._scaler'`

**Comment by PierreRaybaut** [comment]
Those warnings mean that you haven't build the  C++ extensions, or at least that the `plotpy` module that is imported is missing those extensions. Maybe have you run the `cdl` command without having change the current directory from your previous build commands and you are importing the `plotpy` module from the repository instead of the one that you have just installed?

**Comment by marcel-goldschen-ohm** [comment]
Sorry if this should really be asked elsewhere, but I can't figure out based on the docs how to do much other than draw an ROI and see its basic stats. For example, with a signal selected (I just created a sinusoid via the UI) I only have Edit, View, and ? menus, but no menus for Operation, Processing, Computing? How might I fit a function to a curve? Or is this not yet possible?

**Comment by marcel-goldschen-ohm** [comment]
You nailed it! The module not found error is gone if I run cdl from another directory!

But now I only have Edit and ? menus with a signal selected.

**Comment by PierreRaybaut** [comment]
> You nailed it! The module not found error is gone if I run cdl from another directory!
> 
> But now I only have Edit and ? menus with a signal selected.

Could please drag and drop a screenshot?

The fact that some of the menus are not shown is clearly not normal. I'm not sure why it's happening.

**Comment by PierreRaybaut** [comment]
> > You nailed it! The module not found error is gone if I run cdl from another directory!
> > 
> > But now I only have Edit and ? menus with a signal selected.
> 
> Could please drag and drop a screenshot?
> 
> The fact that some of the menus are not shown is clearly not normal. I'm not sure why it's happening.

Are there any traceback in the console?

**Comment by PierreRaybaut** [comment]
> > > You nailed it! The module not found error is gone if I run cdl from another directory!
> > > 
> > > But now I only have Edit and ? menus with a signal selected.
> > 
> > Could please drag and drop a screenshot?
> > 
> > The fact that some of the menus are not shown is clearly not normal. I'm not sure why it's happening.
> 
> Any traceback in the console?

Also, you may try and add a "DEBUG" environment variable and set it to "1", for a time being: this will redirect stdin/stout to the terminal (instead of the embedded Qt comsole).

**Comment by marcel-goldschen-ohm** [comment]
Only have Edit and ? menus with a signal selected.

<img width="987" alt="Screen Shot 2023-12-21 at 3 33 31 PM" src="https://github.com/Codra-Ingenierie-Informatique/DataLab/assets/7348010/fb4120c4-a14a-411f-a463-f7ed16e17b23">

**Comment by PierreRaybaut** [comment]
> Only have Edit and ? menus with a signal selected.
> 
> <img width="987" alt="Screen Shot 2023-12-21 at 3 33 31 PM" src="https://github.com/Codra-Ingenierie-Informatique/DataLab/assets/7348010/fb4120c4-a14a-411f-a463-f7ed16e17b23">
> 

Thanks a lot for the screenshot.
That's really unexpected. I assume it's specific to the MacOS platform as we've thoroughly tested DataLab on both Windows and Linux (at least CentOS and Ubuntu).

I'm still interested in any traceback or warning you may have, with or without the DEBUG environment variable.

Plus, if you could share your Python install details, it surely could help understanding what's going on (e.g. the output of `pip list`).

**Comment by PierreRaybaut** [comment]
I have found some interesting leads on how to solve this missing menu issue (and I confirm that it seems to be a MacOS specific issue).

Some things could be done easily, and if I can come up with something, I'll add a fix with the next release (scheduled for tomorrow if everything's ok).

**Comment by marcel-goldschen-ohm** [comment]
Odd deprecation warning in output of `pip list`:

```
DEPRECATION: Loading egg at XXX/miniconda3/envs/datalab/lib/python3.12/site-packages/PlotPy-2.0.2-py3.12-macosx-10.9-x86_64.egg is deprecated. pip 24.3 will enforce this behaviour change. A possible replacement is to use pip for package installation.. Discussion can be found at https://github.com/pypa/pip/issues/12330
DEPRECATION: Loading egg at XXX/miniconda3/envs/datalab/lib/python3.12/site-packages/tifffile-2023.12.9-py3.12.egg is deprecated. pip 24.3 will enforce this behaviour change. A possible replacement is to use pip for package installation.. Discussion can be found at https://github.com/pypa/pip/issues/12330
Package                       Version
----------------------------- ----------
aiohttp                       3.9.1
aiosignal                     1.3.1
alabaster                     0.7.13
astroid                       3.0.2
attrs                         23.1.0
Babel                         2.14.0
black                         23.12.0
cdl                           0.9.2
certifi                       2023.11.17
charset-normalizer            3.3.2
click                         8.1.7
coverage                      7.3.4
Cython                        3.0.7
dill                          0.3.7
docutils                      0.20.1
frozenlist                    1.4.1
guidata                       3.2.2
h5py                          3.10.0
idna                          3.6
imageio                       2.33.1
imagesize                     1.4.1
iniconfig                     2.0.0
isort                         5.13.2
Jinja2                        3.1.2
lazy_loader                   0.3
markdown-it-py                3.0.0
MarkupSafe                    2.1.3
mccabe                        0.7.0
mdit-py-plugins               0.4.0
mdurl                         0.1.2
multidict                     6.0.4
mypy-extensions               1.0.0
myst-parser                   2.0.0
networkx                      3.2.1
numpy                         1.26.2
opencv-python                 4.8.1.78
packaging                     23.2
pathspec                      0.12.1
Pillow                        10.1.0
pip                           23.3.2
platformdirs                  4.1.0
PlotPy                        2.0.2
PlotPy                        2.0.2
pluggy                        1.3.0
psutil                        5.9.7
Pygments                      2.17.2
pylint                        3.0.3
PyQt5                         5.15.10
PyQt5-Qt5                     5.15.11
PyQt5-sip                     12.13.0
pytest                        7.4.3
pytest-cov                    4.1.0
pytest-qt                     4.2.0
pytest-xvfb                   3.0.0
python-docs-theme             2023.9
PythonQwt                     0.11.2
PyVirtualDisplay              3.0
PyWavelets                    1.5.0
PyYAML                        6.0.1
QtPy                          2.4.1
requests                      2.31.0
scikit-image                  0.22.0
scipy                         1.11.4
setuptools                    68.2.2
snowballstemmer               2.2.0
Sphinx                        7.2.6
sphinx-copybutton             0.5.2
sphinx-qt-documentation       0.4.1
sphinxcontrib-applehelp       1.0.7
sphinxcontrib-devhelp         1.0.5
sphinxcontrib-htmlhelp        2.0.4
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.6
sphinxcontrib-serializinghtml 1.1.9
tifffile                      2023.12.9
tifffile                      2023.12.9
tomli                         2.0.1
tomlkit                       0.12.3
urllib3                       2.1.0
wheel                         0.42.0
yarl                          1.9.4
```

**Comment by marcel-goldschen-ohm** [comment]
Console output when running `cdl` with env DEBUG=1:
```
*** DEBUG mode *** [Reset configuration file, do not redirect std I/O]
<cdl.core.gui.main.CDLMainWindow object at 0x1374e7410>: Starting initialization
<cdl.core.gui.main.CDLMainWindow object at 0x1374e7410>: 2 plugin(s) found
<class 'cdl.plugins.PluginRegistry'>: Plugin Test data registered
qt.qpa.fonts: Populating font family aliases took 227 ms. Replace uses of missing font family "Default" with one that exists to avoid this cost. 
*** DEBUG mode *** [Reset configuration file, do not redirect std I/O]
<cdl.core.gui.main.CDLMainWindow object at 0x1374e7410>: Initialization done
*** DEBUG mode *** [Reset configuration file, do not redirect std I/O]
```

**Comment by marcel-goldschen-ohm** [comment]
Thanks @PierreRaybaut for looking into this so quickly. If you find a simple fix for MacOS that would be great!

**Comment by PierreRaybaut** [comment]
Part 1. of the issue is now fixed with [PlotPy v2.0.3](https://github.com/PlotPyStack/PlotPy/releases/tag/v2.0.3).

**Comment by PierreRaybaut** [comment]
So, this should now be fixed in DataLab v0.10.
I will release this new version in a few minutes/hours.

**Comment by PierreRaybaut** [comment]
Part 2. of the issue has been fixed by 384296f6d6ca50ba1a8a08aab56eb23ae63962fa.

DataLab v0.10.0 includes this fix.
I'll let the issue opened until you can confirm this is really fixed on your side (I was not able to test the fix on MacOS).

**Comment by marcel-goldschen-ohm** [comment]
Wow, thanks for the incredibly fast fix!
I confirm that now `pip install PyQt5 cdl; cdl` launches UI with all menus available.

P.S. I see options for a few fits such as Gaussian, etc. But is there a generic curve fitting UI, e.g. for a custom equation? Would be nice to have a spline fit as well. Or do I need to just implement those via a plugin?

**Comment by PierreRaybaut** [comment]
> Wow, thanks for the incredibly fast fix! I confirm that now `pip install PyQt5 cdl; cdl` launches UI with all menus available on MacOS 12.6.9.

Great news, thanks for testing!

> P.S. I see options for a few fits such as Gaussian, etc. But is there a generic curve fitting UI, e.g. for a custom equation? Would be nice to have a spline fit as well. Or do I need to just implement those via a plugin?

Currently, there is no generic curve fitting UI but that may be implemented in the near future: please do not hesitate to create a "Feature request" [here](https://github.com/Codra-Ingenierie-Informatique/DataLab/issues/new/choose).

Same for the spline fit: please do not hesitate to create a "Feature request" [here](https://github.com/Codra-Ingenierie-Informatique/DataLab/issues/new/choose).

Or, you may also implement those via a plugin, indeed.
Even better, you may contribute for this feature to be available in the main application.
